src_zlib/service/zlib_service.cpp

```diff
@@ -214,6 +234,163 @@ class probe
 
 //------------------------------------------------
 
+
+namespace {
+void* zalloc_impl(
+    void* /* opaque */,
+    unsigned items,
+    unsigned size)
+{
+    try {
+        return ::operator new(items * size);
+    } catch(std::bad_alloc const&) {
+        return Z_NULL;
+    }
+}
+
+void zfree_impl(void* /* opaque */, void* addr)
+{
+    ::operator delete(addr);
+}
+} // namespace
+
+class BOOST_HTTP_PROTO_ZLIB_DECL
+    zlib_filter final : public filter
+{
+private:
+    z_stream stream_;
+
+    void init(bool use_gzip);
+
+public:
+    zlib_filter(bool use_gzip = false);
+    ~zlib_filter();
+
+    zlib_filter(zlib_filter const&) = delete;
+    zlib_filter& operator=(zlib_filter const&) = delete;
+
+    filter::results
+    on_process(
+        buffers::mutable_buffer out,
+        buffers::const_buffer in,
+        bool more) override;
+};
+
+zlib_filter::
+zlib_filter(bool use_gzip)
+{
+    stream_.zalloc = &zalloc_impl;
+    stream_.zfree = &zfree_impl;
+    stream_.opaque = nullptr;
+    init(use_gzip);
+}
+
+zlib_filter::
+~zlib_filter()
+{
+    deflateEnd(&stream_);
+}
+
+void
+zlib_filter::
+init(bool use_gzip)
+{
+    int ret = -1;
+
+    int window_bits = 15;
+    if( use_gzip )
+        window_bits += 16;
```

Comment:
the memory probe implementation needs to be adjusted so it measures both values of use_gzip